File: buildtools/parsers/custom.js
/**
*
*   Custom Parser for JavaScript/Node
*
*   @author Nikos M.  
*   https://foo123.github.com/
*   http://nikos-web-development.netai.net/
*
**/
(function(undef){
    
    var  fs = (require) ? require('fs') : null,
        NL = /\n\r|\r\n|\r|\n/g, 
        BLOCK = /^@(([a-zA-Z0-9\-_]+)\s*(=\[\]|=\{\}|=)?)/,
        ENDBLOCK = /^(@\s*)+/,
        MAP = 1, LIST = 2, VAL = 0
    ;
    
    var 
        trim = function(s) {
            return (s) ? s.replace(/^\s+/g, '').replace(/\s+$/g, '') : s;
        },
        
        removeComment = function(s, comm) {
            s = s.split( comm );
            return trim( s[0] );
        },
        
        startsWith = function(s, prefix) { return (s && (prefix == s.substr(0, prefix.length))); },
        
        parseStr = function(s, q) {
            var endq = s.indexOf(q, 1);
            var quoted = s.substr(1, endq-1);
            var rem = trim( s.substr(endq) );
            return [quoted, rem];
        },
        
        getQuotedValue = function( line ) {
            var linestartswith = line[0];
            
            // quoted string
            if ( '"'==linestartswith || "'"==linestartswith || "`"==linestartswith )
            {
                var res = parseStr( line, linestartswith );
                return res[0];
            }
            // un-quoted string
            else
            {
                return trim( line );
            }
        },
        
        getKeyValuePair = function( line ) {
            var linestartswith = line[0];
            
            // quoted string
            if ( '"'==linestartswith || "'"==linestartswith || "`"==linestartswith )
            {
                var res = parseStr(line, linestartswith);
                var key = res[0];
                line = res[1];
                
                // key-value pair
                if ( line.indexOf('=')>-1 )
                {
                    line = line.split('=');
                    line.shift()
                    value = line.join('=');
                    if ( startsWith(value, "[]"))
                    {
                        return [key, [], LIST];
                    }
                    else if ( startsWith(value, "{}"))
                    {
                        return [key, {}, MAP];
                    }
                    
                    if ( value )
                    {
                        value = trim(value);
                        var valuestartswith = value[0];
                        
                        // quoted value
                        if ( '"'==valuestartswith || "'"==valuestartswith || "`"==valuestartswith )
                        {
                            res = parseStr(value, valuestartswith);
                            value = res[0];
                        }
                    }
                    return [key, value, VAL];
                }
            }
            // un-quoted string
            else
            {
                line = line.split('=');
                
                var key = trim(line.shift());
                var value = line.join('=');
                
                if ( startsWith(value, "[]"))
                {
                    return [key, [], LIST];
                }
                else if ( startsWith(value, "{}"))
                {
                    return [key, {}, MAP];
                }
                
                if ( value )
                {
                    value = trim(value);
                    var valuestartswith = value[0];
                    
                    // quoted value
                    if ( '"'==valuestartswith || "'"==valuestartswith || "`"==valuestartswith )
                    {
                        var res = parseStr(value, valuestartswith);
                        value = res[0];
                    }
                }
                
                return [key, value, VAL];
            }
        }
    ;
    
    var CustomParser = self = {
        
        fromString : function(s)  {
            
            // settings buffer
            var settings = {};
            
            // current settings options
            var currentBuffer = settings,
                currentBlock = null, 
                currentPath = [],
                isType = VAL
            ;

            // parse the lines
            var i, line, lines, lenlines, block, endblock, j, jlen, numEnds, keyval;

            s = ''+s;
            lines = s.split( NL );
            lenlines = lines.length;
            
            // parse it line-by-line
            for (i=0; i<lenlines; i++)
            {
                // strip the line of comments and extra spaces
                line = removeComment( lines[i], "#" );

                // comment or empty line, skip it
                if ( !line.length ) continue;

                // block/directive line, parse it
                if ( startsWith(line, '@') )
                {
                    block = BLOCK.exec( line );
                    endblock = ENDBLOCK.exec( line );
                    
                    if ( block )
                    {
                        currentBlock = block[2];
                        if ( !block[3] || '='==block[3] ) isType = VAL;
                        else if ( '=[]'==block[3] ) isType = LIST;
                        else if ( '={}'==block[3] ) isType = MAP;
                        
                        currentPath.push( [currentBlock, isType] );
                        if ( currentPath.length>1 )
                        {
                            currentBuffer = settings;
                            for (j=0, l1=currentPath.length-1; j<l1; j++)
                            {
                                currentBuffer = currentBuffer[ currentPath[j][0] ];
                            }
                        }
                        if ( undef === currentBuffer[ currentBlock ] )
                        {
                            if (LIST == isType)
                                currentBuffer[ currentBlock ] = [];
                            else if (MAP == isType)
                                currentBuffer[ currentBlock ] = {};
                            else
                                currentBuffer[ currentBlock ] = '';
                        }
                    }
                    
                    else if ( endblock )
                    {
                        numEnds = line.split("@").length-1;
                        
                        for (j=0; j<numEnds; j++)
                            currentPath.pop();
                        
                        currentBuffer = settings;
                        if ( currentPath.length > 0 )
                        {
                            if ( currentPath.length > 1 )
                            {
                                for (j=0, jlen=currentPath.length-1; j<jlen; j++)
                                {
                                    currentBuffer = currentBuffer[ currentPath[j][0] ];
                                }
                            }
                            currentBlock = currentPath[ currentPath.length-1 ][0];
                            isType = currentPath[ currentPath.length-1 ][1];
                        }
                        else
                        {
                            currentBlock = null;
                            isType = VAL;
                        }
                    }
                    
                    continue;
                }
                
                // if any settings need to be stored, store them in the appropriate buffer
                if ( currentBlock && currentBuffer )  
                {
                    if ( MAP == isType )
                    {
                        keyval = getKeyValuePair( line );
                        
                        currentBuffer[ currentBlock ][ keyval[0] ] = keyval[1];
                        
                        if ( LIST == keyval[2] || MAP == keyval[2] )
                        {
                            currentPath.push( [keyval[0], keyval[2]] );
                            currentBuffer = currentBuffer[ currentBlock ];
                            currentBlock = keyval[0];
                            isType = keyval[2];
                        }
                    }
                    else if ( LIST == isType )
                    {
                        currentBuffer[ currentBlock ].push( getQuotedValue( line ) );
                    }
                    else //if ( VAL == isType )
                    {
                        currentBuffer[ currentBlock ] = getQuotedValue( line );
                        currentBlock = null;
                        isType = VAL;
                        
                        /*
                        currentPath.pop();
                        currentBuffer = settings;
                        if ( currentPath.length > 0 )
                        {
                            if ( currentPath.length > 1 )
                            {
                                for (j=0, jlen=currentPath.length-1; j<jlen; j++)
                                {
                                    currentBuffer = currentBuffer[ currentPath[j][0] ];
                                }
                            }
                            currentBlock = currentPath[ currentPath.length-1 ][0];
                            isType = currentPath[ currentPath.length-1 ][1];
                        }
                        else
                        {
                            currentBlock = null;
                            isType = VAL;
                        }
                        */
                    }
                }
            }
            
            return settings;
        },
        
        fromFile : function(filename, keysList, rootSection) {
            if (fs)
            {
                return self.fromString( fs.readFileSync(filename) );
            }
            return '';
        }
    };
    
    // export it
    if ('undefined' != typeof (module) && module.exports)  module.exports = CustomParser;
    
    else if ('undefined' != typeof (exports)) exports = CustomParser;
    
    else this.CustomParser = CustomParser;
    
}).call(this);

File: buildtools/parsers/custom.py
##
#
#   Custom Parser for Python 2.x, 3.x
#
#   @author Nikos M.  
#   https://foo123.github.com/
#   http://nikos-web-development.netai.net/
#
##
#import pprint
import re

class CustomParser():
    """Custom parser for Python"""
    
    MAP = 1
    LIST = 2
    VAL = 0
    
    NL = None
    BLOCK = None
    ENDBLOCK = None
    
    def removeComment(s, comm):
        s = s.split( comm )
        return s[0].strip()
    
    
    def parseStr(s, q):
        endq = s.find(q, 1)
        quoted = s[1:endq]
        rem = s[endq+1:].strip()
        
        return quoted, rem
    
    def getQuotedValue( line ):
        _self = CustomParser
        
        linestartswith = line[0]
        
        # quoted string
        if '"'==linestartswith or "'"==linestartswith or "`"==linestartswith:
        
            key, line = _self.parseStr(line, linestartswith)
            return key
        
        # un-quoted string
        else:
            return line.strip()
        
    
    
    def getKeyValuePair(line):
        _self = CustomParser
        
        linestartswith = line[0]
        # quoted string
        if '"' == linestartswith or "'" == linestartswith or "`" == linestartswith:
        
            key, line = _self.parseStr(line, linestartswith)
            
            # key-value pair
            if line.find('=', 0)>-1:
                
                line = line.split('=')
                line.pop(0)
                value = "=".join(line)
                
                if value.startswith("[]"):
                
                    return [key, [], _self.LIST]
                
                elif value.startswith("{}"):
                
                    return [key, {}, _self.MAP]
                
                if value:
                
                    value = value.strip()
                    valuestartswith = value[0]
                    
                    # quoted value
                    if '"'==valuestartswith or "'"==valuestartswith or "`"==valuestartswith:
                        value, rem = _self.parseStr(value, valuestartswith)
                
                return [key, value, _self.VAL]
            
        
        # un-quoted string
        else:
        
            line = line.split('=')
            key = line.pop(0).strip()
            value = "=".join(line)
            
            if value.startswith("[]"):
            
                return [key, [], _self.LIST]
            
            elif value.startswith("{}"):
            
                return [key, {}, _self.MAP]
            
            if value:
                
                value = value.strip()
                valuestartswith = value[0]
                
                # quoted value
                if '"'==valuestartswith or "'"==valuestartswith or "`"==valuestartswith:
                    value, rem = _self.parseStr(value, valuestartswith)
            
            return [key, value, _self.VAL]
    
    
    def fromString(s):
        _self = CustomParser
        
        if not _self.NL:
            _self.NL = re.compile(r'\n\r|\r\n|\r|\n')
        
        if not _self.BLOCK:
            _self.BLOCK = re.compile(r'^@(([a-zA-Z0-9\-_]+)\s*(=\[\]|=\{\}|=)?)')
        
        if not _self.ENDBLOCK:
            _self.ENDBLOCK = re.compile(r'^(@\s*)+')
            
        # settings buffers
        settings = {}
        
        currentBuffer = settings
        currentPath = []
        currentBlock = None
        isType = _self.VAL
        
        # parse the lines
        lines = re.split(_self.NL, str(s))
        
        # parse it line-by-line
        for line in lines:
            
            # strip the line of comments and extra spaces
            line = _self.removeComment(line, "#")
            
            # comment or empty line, skip it
            if 0==len(line): continue
            
            # block/directive line, parse it
            if line.startswith('@'):
                
                block = _self.BLOCK.match( line )
                endblock = _self.ENDBLOCK.match( line )
                
                if block:
                
                    currentBlock = block.group(2)
                    block3 = block.group(3)
                    if  (block3 is None) or ('='==block3): isType = _self.VAL
                    elif '=[]'==block3: isType = _self.LIST
                    elif '={}'==block3: isType = _self.MAP
                    
                    currentPath.append( [currentBlock, isType] )
                    currLen = len(currentPath)
                    if currLen>1:
                    
                        currentBuffer = settings
                        for j in range(currLen-1):
                        
                            currentBuffer = currentBuffer[ currentPath[j][0] ]
                        
                    
                    if currentBlock not in currentBuffer:
                    
                        if _self.LIST == isType:
                            currentBuffer[ currentBlock ] = []
                        elif _self.MAP == isType:
                            currentBuffer[ currentBlock ] = {}
                        else:
                            currentBuffer[ currentBlock ] = ''
                    
                
                
                elif endblock:
                
                    numEnds = len(line.split("@"))-1
                    
                    for j in range(numEnds):
                        if len(currentPath): currentPath.pop()
                        else: break
                    
                    currentBuffer = settings
                    currLen = len(currentPath)
                    if currLen > 0:
                    
                        if currLen > 1:
                        
                            for j in range(currLen-1):
                            
                                currentBuffer = currentBuffer[ currentPath[j][0] ]
                            
                        
                        currentBlock = currentPath[ currLen-1 ][0]
                        isType = currentPath[ currLen-1 ][1]
                    
                    else:
                    
                        currentBlock = None
                        isType = _self.VAL
                    
                
                continue
            
            # if any settings need to be stored, store them in the appropriate buffer
            if (currentBlock is not None) and (currentBuffer is not None): 
                if _self.MAP == isType:
                
                    keyval = _self.getKeyValuePair( line )
                    
                    currentBuffer[ currentBlock ][ keyval[0] ] = keyval[1]
                    
                    if _self.LIST == keyval[2] or _self.MAP == keyval[2]:
                    
                        currentPath.append( [keyval[0], keyval[2]] )
                        currentBuffer = currentBuffer[ currentBlock ]
                        currentBlock = keyval[0]
                        isType = keyval[2]
                    
                
                elif _self.LIST == isType:
                
                    currentBuffer[ currentBlock ].append( _self.getQuotedValue( line ) )
                
                else: # elif _self.VAL == isType:
                
                    currentBuffer[ currentBlock ] = _self.getQuotedValue( line )
                    currentBlock = None
                    isType = _self.VAL
                    
                    #if len(currentPath): currentPath.pop()
                    #currentBuffer = settings
                    #currLen = len(currentPath)
                    #if currLen > 0:
                    #
                    #    if currLen > 1:
                    #    
                    #        for j in range(currLen-1):
                    #        
                    #            currentBuffer = currentBuffer[ currentPath[j][0] ]
                    #        
                    #    
                    #    currentBlock = currentPath[ currLen-1 ][0]
                    #    isType = currentPath[ currLen-1 ][1]
                    #
                    #else:
                    #
                    #    currentBlock = None
                    #    isType = _self.VAL
                
        
        return settings

    
    def fromFile(filename):
        s = ''
        with open(filename, 'r') as f:  s = f.read()
        return CustomParser.fromString(s)
        


            
# for use with 'import *'
__all__ = [ 'CustomParser' ]
